tls/s2n_handshake_io.c

```diff
@@ -1022,12 +1006,37 @@ static int handshake_read_io(struct s2n_connection *conn)
     return 0;
 }
 
-static int s2n_try_delete_session_cache(struct s2n_connection *conn) 
+static int s2n_handle_retry_state(struct s2n_connection *conn)
 {
-    notnull_check(conn);
-    if (s2n_errno != S2N_ERR_BLOCKED && s2n_allowed_to_cache_connection(conn) && conn->session_id_len) {
-        conn->config->cache_delete(conn, conn->config->cache_delete_data, conn->session_id, conn->session_id_len);
+    /* If we were blocked reading a record, then the handler is waiting on external data.
+     * The handler will know how to continue, so we should call the handler right away.
+     * We aren't going to read more handshake data yet because the current message has
+     * not finished processing. */
+    int r = ACTIVE_STATE(conn).handler[conn->mode] (conn);
+
+    if (r < 0) {
+        /* If the handler is still waiting for data, return control to the caller. */
+        if (s2n_errno == S2N_ERR_BLOCKED) {
+            S2N_ERROR_PRESERVE_ERRNO();
+        }
+
+        /* Otherwise there is some other problem and we should kill the connection. */
+        if (s2n_allowed_to_cache_connection(conn) && conn->session_id_len) {
```

Comment:
switch these.

checking `conn->session_id_len` doesn't require creating a stack frame, so can short cut that if it's not set.

Reply:
Updated.

Reply:
Added a callback return code, S2N_CALLBACK_BLOCKED which is -2.